### PR TITLE
Fix RtlUnwind usage from CRT wine code

### PR DESCRIPTION
**Pull request by tkreuzer**
## Purpose

Fixes regression in C++ exception handling.

JIRA issue: [CORE-19392](https://jira.reactos.org/browse/CORE-19392), [CORE-19397](https://jira.reactos.org/browse/CORE-19397)

## Proposed changes

- Add apitest for x86 RtlUnwind
- Add asm wrapper around RtlUnwind for wine code

## Test
- x86 KVM: https://reactos.org/testman/compare.php?ids=92673,92674,92677
- WHS: https://reactos.org/testman/compare.php?ids=92730

## TODO

- [x] Run on testbot (wip)
- [x] Verification that both issues are fixed

### CODE REVIEW

**Comment by learn-more** [comment]
I might be missing something, but I do not see this new `__wine__RtlUnwind` used anywhere in this PR?

File: sdk/lib/crt/wine/rosglue_i386.s [hunk 1]
```diff
@@ -0,0 +1,47 @@
+
+#include <asm.inc>
+
+.code
+
+EXTERN _RtlUnwind@16:PROC
+
+// ASM wrapper for Wine code. This is needed, because Wine code expects
+// RtlUnwind to restore the non-volatile registers, before returning, but
+// ours / the native one does not do that.
```
Comment (HBelusca) [inline sdk/lib/crt/wine/rosglue_i386.s:10]: "the native one" == Windows' one? Is this true for all Windows versions?

**Comment by tkreuzer** [comment]
> I might be missing something, but I do not see this new `__wine__RtlUnwind` used anywhere in this PR?

You have to look harder :)
`#define RtlUnwind __wine__RtlUnwind`

**Review by learn-more — APPROVED** [review]
> > I might be missing something, but I do not see this new `__wine__RtlUnwind` used anywhere in this PR?
> 
> You have to look harder :) `#define RtlUnwind __wine__RtlUnwind`

Ah yes, thanks!

I could not spot any obvious flaws, just one minor nitpick on the comments

File: sdk/lib/crt/wine/rosglue_i386.s (hunk 1 above)
Comment (tkreuzer) [reply sdk/lib/crt/wine/rosglue_i386.s:10]: It's true for Win 2003 (see WHS test linked in the description) and Windows 10 WOW64.

File: modules/rostests/apitests/rtl/i386/RtlUnwind-asm.s [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    mov ecx, _g_InContext[CONTEXT_ECX]
+    mov edx, _g_InContext[CONTEXT_EDX]
+    mov esi, _g_InContext[CONTEXT_ESI]
+    mov edi, _g_InContext[CONTEXT_EDI]
+
+    /* Call the native function */
+    push dword ptr [ebp + 20] // retval
+    push dword ptr [ebp + 16] // pRecord
+    push dword ptr [ebp + 12] // targetIp
+    push dword ptr [ebp + 8] // pEndFrame
```
[the remark this answers is not in the document]
Comment (tkreuzer) [reply modules/rostests/apitests/rtl/i386/RtlUnwind-asm.s:49]: Yeah, I can change them to the ones we use.

File: sdk/lib/crt/wine/rosglue_i386.s (hunk 1 above)
Comment (HBelusca) [reply sdk/lib/crt/wine/rosglue_i386.s:10]: ok, "wow64" because the rtlunwind is x86 only?

**Review by HBelusca — APPROVED** [review]

File: sdk/lib/crt/wine/rosglue_i386.s (hunk 1 above)
Comment (tkreuzer) [reply sdk/lib/crt/wine/rosglue_i386.s:10]: x64 also has RtlUnwind, but that works differently. The issue here is x86 specific.

**Pull request merged**